First version of retrieval tutorial (#1)
nasa-fornax/fornax-howtos: tutorials/exoplanet_retrievals/quickstart.md

Many retrieval tools (including pRT) require local opacity files that are too large to package in this repository.

- You must download/install opacities in your own Fornax storage.
- Opacity files are **not** distributed in this repo due to size constraints.
- pRT opacity installation instructions: <https://petitradtrans.readthedocs.io/en/latest/content/available_opacities.html>

If centralized opacity hosting on Fornax would help your workflow, please share feedback with the Fornax team.

+++

## 3. Run retrieval scripts with MPI

For pRT workflows that use `pymultinest`, write retrieval wrappers as Python scripts and launch with `mpirun`.

This repository includes:

- `examples/exoplanet-retrievals/run_prt_basic_pymultinest_mpi.py`
- `examples/exoplanet-retrievals/hst_example_clear_spec.txt`

Example MPI launch commands:

```bash
# Small validation run (recommended first test)
mpirun -np 2 python examples/exoplanet-retrievals/run_prt_basic_pymultinest_mpi.py --use-mpi --n-live-points 40

# Medium run for quick scaling checks (assuming you have ncores > 8)
mpirun -np 8 python examples/exoplanet-retrievals/run_prt_basic_pymultinest_mpi.py --use-mpi --n-live-points 40

> ```

Reviewer: Briefly explain what the user should expect to happen when these commands are run.

Author: good idea.

> 

Author: ```suggestion
Expected output:

When you run the retrieval with mpirun, you should see:
- Validation checks confirming that the input data and model functions execute without errors.
- A MultiNest header showing the number of live points and model dimensionality.
- Periodic sampling updates (acceptance rate, total samples, and evidence ln Z estimates).
- A final summary including log evidence (ln Z ± uncertainty), total likelihood evaluations, and retrieved parameter values with uncertainties

For the small validation run (--n-live-points 40), expect on the order of ~1,000–2,000 likelihood evaluations and a runtime of a few minutes, depending on the number of MPI processes used. 
Larger n_live_points values will increase runtime substantially but produce more stable evidence estimates and posterior constraints.

Output files are written to:
`retrievals/runs/out_PMN/`
This directory contains the MultiNest chain files, posterior summaries, and generated diagnostic plots (including corner plots). 
After the run completes, check this directory for retrieval results and figures.
```

### Benchmark figure 1: speedup vs nprocs

```{figure} ../examples/exoplanet-retrievals/speedup_vs_nprocs.png
:alt: Speedup versus number of MPI processes
:name: fig-speedup-vs-nprocs

Speedup is defined as the wall time for 1 processor divided by the wall time for *nprocs* processors. Typical behavior includes (1) non-linear scaling, (2) useful speedup up to about 0.5-0.75× of available cores, and (3) larger gains for larger `n_live_points`, which is more representative of realistic retrieval configurations. Overheads are a major cause of non-linear speedups and limits in gains as a function of `nprocs`.
```

### Benchmark figure 2: seconds per likelihood evaluation vs nprocs

```{figure} ../examples/exoplanet-retrievals/seconds_per_likelihood_eval_vs_nprocs.png
:alt: Seconds per likelihood evaluation versus number of MPI processes
:name: fig-seconds-per-like

A benchmarking curve like this for your specific retrieval setup (including your opacity selection, clouds, and model complexity) can help predict full runtime and guide resource selection.

> ```

Reviewer: These will probably look fine when rendered for a website or by jupyter but they don't give a nice experience when viewed as a regular markdown preview:

<img width="1160" height="403" alt="Image" src="https://github.com/user-attachments/assets/a76f92b5-ddfb-46f9-b0a9-121265230c39" />

By contrast, here's an example using standard markdown syntax (left) and what that looks like when viewed thru a standard markdown preview (right):

<img width="1609" height="1039" alt="Image" src="https://github.com/user-attachments/assets/f948e66b-714f-4474-877c-4d1ea60b2843" />

I think we should use the standard markdown syntax here since it's not uncommon for users to be looking at a standard markdown preview rather than the website or jupyter. (The situation is different in our documentation repos because there we only ever expect users to be looking at the rendered website, not the files.)

Author: I know there was discussion about this for some recent tutorials, I thought the result of that was to use the special formatting.  If not, I can change it.  This is sort of a doc in the sense that no-one needs to run it.

Reviewer: I thought the result of that discussion was that in notebooks repos we should use the syntax that works everywhere. Sounds like I overestimated the amount of consensus we had.

Reviewer 2: why do we expect anyone to look at the notebooks with anything but jupyter or in the rendering?

Reviewer 2: (I mean we don't expect them to have a nice experience looking at the json notebooks with a terminal text editor, so this expectation doesn't feel fair -- and as a result would stop us from using nice myst features)

Reviewer: I expect that some people will look at the notebooks on github. Also that some people will clone the repo and be working with the markdown files without using jupyter. In those cases, myst syntax means that images don't show up, caption text doesn't wrap, and internal links are totally broken. By contrast, standard markdown syntax works everywhere. The tradeoff is that extra myst features wouldn't be used in jupyter and rendering, as you say, but at least nothing is broken. What are the extra myst features we'd be missing out on? Thumbnail previews on hover over links is one that I'm aware of.

Author: I will leave this as is for now.  If the team has a discussion on this topic and wants to revise the general plan about using myst features, then we can open an issue to make changes.  I don't want this PR held up waiting for the outcome of that discussion.

- First-time environment creation may take more than 10 minutes.
- Environments containing packages to run retrievals can consume significant storage.

### 1.2 Activate the environment:

From the terminal, activate the environment with the following command:
```bash
micromamba activate $USER_ENV_DIR/exo
```

+++

## 2. Install opacity files (required and large)

Many retrieval tools (including pRT) require local opacity files that are too large to package in this repository.
Opacity files are **not** distributed in this repo due to size constraints.

```{admonition} Fornax note
pRT supports automatic opacity downloads via [Keeper](https://keeper.mpdl.mpg.de/d/ccf25082fda448c8a0d0), but this method is not available on Fornax.
Follow the download steps in section 2.2 below instead.
On different compute resources, the automatic Keeper download may work for you.
```

### 2.1 Opacities required for this tutorial

This tutorial uses the following opacity files:

- **Line species (correlated-k):** `H2O__POKAZATEL`, `CH4__HITEMP`, `CO-NatAbund__HITEMP`
- **Collision-induced absorption (CIA):** `H2--H2`, `H2--He`
- **Rayleigh scattering:** `H2`, `He` (built into pRT — no download needed)

> Opacity files are stored under an `input_data/` directory alongside the retrieval script in `examples/exoplanet-retrievals/input_data/`.

Reviewer: ```suggestion
Opacity files are stored under an `input_data/` directory alongside the retrieval script in `tutorials/exoplanet_retrievals/input_data/`.
```

This path is set automatically when you run the setup script described below.

### 2.2 Download opacities using the setup script

This repository includes a setup script that creates the `input_data/` directory, registers it with pRT, and downloads the required opacity files.

```{admonition} Note
:class: note
The setup script permanently updates your global pRT configuration (`~/.petitradtrans/petitradtrans_config_file.ini`) to point to the tutorial `input_data/` directory.
If you already have pRT configured with your own opacity data elsewhere, you can restore your original path afterwards by calling `petitradtrans_config_parser.set_input_data_path('/your/original/path')`.
```

```bash

> python examples/exoplanet-retrievals/setup_opacities.py

Reviewer: I suggest we assume the user is in the same directory as this notebook. That's where they have to be to use the included conda file in the previous section.

```suggestion
python setup_opacities.py
```

```

```{admonition} Warning
:class: warning
Opacity files are large — the full set for this tutorial totals approximately 650 MB.
Downloads may take significant time.
Run the setup script in a terminal rather than a notebook to avoid session timeouts.
```

The script triggers pRT's built-in download mechanism by initializing a `Radtrans` object with the exact species used in this tutorial.
If prompted to choose among multiple matching files, type the integer for the `R1000_0.1-250mu` file and press Enter.

If centralized opacity hosting on Fornax would help your workflow, please share feedback with the Fornax team.

+++

## 3. Run retrieval scripts with MPI

For pRT workflows that use `pymultinest`, write retrieval wrappers as Python scripts and launch with `mpirun`.

This repository includes:

- `examples/exoplanet-retrievals/run_prt_basic_pymultinest_mpi.py`

> - `examples/exoplanet-retrievals/hst_example_clear_spec.txt`

Reviewer: ```suggestion
- `tutorials/exoplanet_retrievals/run_retrieval.py`
- `tutorials/exoplanet_retrievals/hst_example_clear_spec.txt`
```

> - `examples/exoplanet-retrievals/hst_example_clear_spec.txt`

Reviewer: ```suggestion
- `tutorials/exoplanet_retrievals/run_retrieval.py`
- `tutorials/exoplanet_retrievals/hst_example_clear_spec.txt`
```

Example MPI launch commands:
Note: The --n-live-points argument sets the number of “live” samples (n_live in MultiNest terminology) maintained during nested sampling; larger values improve the stability of evidence and posterior estimates but increase runtime.
```bash
# Small validation run (recommended first test)
mpirun -np 2 python examples/exoplanet-retrievals/run_prt_basic_pymultinest_mpi.py --use-mpi --n-live-points 40

# Medium run for quick scaling checks (assuming you have ncores > 8)

> mpirun -np 8 python examples/exoplanet-retrievals/run_prt_basic_pymultinest_mpi.py --use-mpi --n-live-points 40

Reviewer: ```suggestion
# Small validation run (recommended first test)
mpirun -np 2 python run_retrieval.py --use-mpi --n-live-points 40

# Medium run for quick scaling checks (assuming you have ncores > 8)
mpirun -np 8 python run_retrieval.py --use-mpi --n-live-points 40
```

```
### 3.1 Expected output

When you run the retrieval with mpirun, you should see:
- Validation checks confirming that the input data and model functions execute without errors.
- A MultiNest header showing the number of live points and model dimensionality.
- Periodic sampling updates (acceptance rate, total samples, and evidence ln Z estimates).
- A final summary including log evidence (ln Z ± uncertainty), total likelihood evaluations, and retrieved parameter values with uncertainties

> For the small validation run (--n-live-points 40), expect on the order of ~1,000–2,000 likelihood evaluations and a runtime of a few minutes, depending on the number of MPI processes used. 

Reviewer: Both small and medium commands use `--n-live-points 40`. I wonder if this line should say (`-np 2`) or if the medium command above should use a different `--n-live-points`?

Larger n_live_points values will increase runtime substantially but produce more stable evidence estimates and posterior constraints.

Output files are written to:
`retrievals/runs/out_PMN/`
This directory contains the MultiNest chain files, posterior summaries, and generated diagnostic plots (including corner plots). 

> After the run completes, check this directory for retrieval results and figures.

Reviewer: I don't see any plots or figures. Here are the files I get:

```
$ ls retrievals/runs/out_PMN/
hst_example_clear_spec_.txt         hst_example_clear_spec_phys_live.points
hst_example_clear_spec_IS.iterinfo  hst_example_clear_spec_post_equal_weights.dat
hst_example_clear_spec_IS.points    hst_example_clear_spec_resume.dat
hst_example_clear_spec_IS.ptprob    hst_example_clear_spec_stats.dat
hst_example_clear_spec_ev.dat       hst_example_clear_spec_stats.json
hst_example_clear_spec_live.points  hst_example_clear_spec_summary.txt
hst_example_clear_spec_params.json
```

I can't open hst_example_clear_spec_IS.iterinfo, hst_example_clear_spec_IS.points, or hst_example_clear_spec_IS.ptprob, so I'm not sure what's in there.

### 3.2 Important Fornax MPI subtleties

When running pRT retrievals with MPI on Fornax, using all available CPU cores can exhaust limited shared-memory space (`/dev/shm`) used for inter-process communication. 
This can fail jobs even when system RAM is still available.

In practice:

1. Speedup is usually sub-linear.
2. Performance gains often flatten before you reach all available cores.
3. Many workflows perform best around roughly half to three-quarters of available cores.

+++

## 4. Benchmark to estimate full retrieval runtime

Full retrieval wall time is difficult to predict in advance. 
Benchmarking your own setup (opacities, chemistry assumptions, cloud setup, live points) helps you choose compute size and estimate total runtime. 
If you anticipate that your retrieval will run longer than a few hours, enabling the [JupyterLab keep-alive feature](https://docs.fornax.sciencecloud.nasa.gov/jupyterlab/#jupyterlab-session-information) can prevent unintended shutdown. 
However, keep-alive sessions incur credit charges for their entire lifetime, regardless of active CPU usage, so plan accordingly and monitor extended runs.

### Benchmark figure 1: speedup vs nprocs

> ```{figure} ../examples/exoplanet-retrievals/speedup_vs_nprocs.png

Reviewer: ```suggestion
```{figure} speedup_vs_nprocs.png
```

:alt: Speedup versus number of MPI processes
:name: fig-speedup-vs-nprocs

Speedup is defined as the wall time for 1 processor divided by the wall time for *nprocs* processors.  
Typical behavior includes (1) non-linear scaling, (2) useful speedup up to about 0.5-0.75× of available cores, and (3) Larger gains are seen for larger n_live_points (shown as n_live in the legend), which controls the number of active samples MultiNest keeps while exploring parameter space. 
Increasing n_live_points improves posterior and evidence stability but increases runtime.
Overheads are a major cause of non-linear speedups and limits in gains as a function of `nprocs`.
```

### Benchmark figure 2: seconds per likelihood evaluation vs nprocs

> ```{figure} ../examples/exoplanet-retrievals/seconds_per_likelihood_eval_vs_nprocs.png

Reviewer: ```suggestion
```{figure} seconds_per_likelihood_eval_vs_nprocs.png
```
